src/ht.c

```diff
@@ -1,239 +1,238 @@
-	SdbHash old_ht = *ht;
-	SdbHashEntry *e;
-	if (!ht || new_size_index >= ARRAY_SIZE (hash_sizes))
-		return;
-	// XXX: This code is redupped! fuck't
-	ht->table = calloc (hash_sizes[new_size_index].size, sizeof (*ht->table));
-	if (!ht->table)
-		return;
-	rehash = 1;
-	ht->size_index = new_size_index;
-	ht->size = hash_sizes[ht->size_index].size;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
-		if (entry_is_present (e))
-			ht_insert (ht, e->hash, e->data, e->iter);
+void ht_free(SdbHash* ht) {
+	ut32 i;
+	for (i = 0; i < ht->size; ++i) {
+		ls_free (ht->table[i]);
 	}
-	free (old_ht.table);
-rehash = 0;
+	free (ht->table);
+	free (ht);
 }
 
-SdbHash* ht_new(SdbListFree f) {
-	SdbHash *ht = R_NEW (SdbHash);
-	if (!ht) return NULL;
-	// TODO: use slices here
-	ht->list = ls_new ();
-	ht->list->free = f;
-	ht->size = hash_sizes[0].size;
-	ht->table = calloc (ht->size, sizeof (*ht->table));
-	if (!ht->table) {
-		free (ht);
-		return NULL;
+// Increases the size of the hashtable by 2.
+static void internal_ht_grow(SdbHash* ht) {
+	SdbHash* ht2;
+	SdbHash swap;
+	SdbKv* kvp;
+	SdbListIter* iter;
+	ut32 i;
+	ut32 sz = ht_primes_sizes[ht->prime_idx];
+	ht2 = internal_ht_new (sz, ht->hashfn, ht->cmp, ht->dupkey,
+			ht->dupvalue, (HtKvFreeFunc)ht->table[0]->free, ht->calcsize);
+	ht2->prime_idx = ht->prime_idx;
+	for (i = 0; i < ht->size; ++i) {
+		ls_foreach (ht->table[i], iter, kvp) {
+			(void)ht_insert (ht2, kvp->key, kvp->value);
+		}
 	}
-	ht->size_index = 0;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	return ht;
+	// And now swap the internals.
+	swap = *ht;
+	*ht = *ht2;
+	*ht2 = swap;
+	ht_free (ht2);
 }
 
-void ht_free(SdbHash *ht) {
-	if (ht) {
-		free (ht->table);
-		ls_free (ht->list);
-		free (ht);
+// Inserts the key value pair key, value into the hashtable.
+// if update is true, allow for updates, otherwise return false if the key
+// already exists.
+static bool internal_ht_insert(SdbHash* ht, bool update, const char* key, const char* value) {
+	SdbKv* kvp;
+	ut32 hash = ht->hashfn (key);
+	ut32 bucket;
+	bool found = true;
+	if (update) {
+		(void)ht_delete_internal (ht, key, &hash);
+	} else {
+		(void)ht_find (ht, key, &found);
+	}
+	if (update || !found) {
+		kvp = calloc (1, sizeof (SdbKv));
+		if (kvp) {
+			if (ht->dupkey) {
+				kvp->key = ht->dupkey (key);
+			} else {
+				kvp->key = (char*)key;
+			}
+			if (ht->dupvalue) {
+				kvp->value = ht->dupvalue (value);
+			} else {
+				kvp->value = (char*)value;
+			}
+			bucket = hash % ht->size;
+			kvp->expire = 0;
+			if (ht->calcsize) {
+				kvp->value_len = (ut32)ht->calcsize (kvp->value);
+			} else {
+				kvp->value_len = 0;
+			}
+			ls_prepend (ht->table[bucket], kvp);
```

Comment:
nope, list wont grow, if inserting the second time, we hit the `ht_delete_internal`